auth_token/contrib/ms_sso/views.py

```diff
@@ -0,0 +1,57 @@
+from django.core.exceptions import ImproperlyConfigured
+from django.contrib import messages
+from django.utils.translation import gettext
+from django.views.generic.base import RedirectView
+from django.contrib.auth import authenticate
+
+from auth_token.utils import login
+
+from .helpers import get_sign_in_flow, acquire_token_by_auth_code_flow
+
+
+class MsLogin(RedirectView):
+
+    def get_redirect_url(self, *args, **kwargs):
+        sign_flow = get_sign_in_flow()
+        sign_flow['next'] = self.request.GET.get('next', '/')
+        if not hasattr(self.request, 'session'):
+            raise ImproperlyConfigured('Django SessionMiddleware must be enabled to use MS SSO')
+
+        self.request.session['auth_token_ms_sso_auth_flow'] = sign_flow
+        return sign_flow['auth_uri']
+
+
+class MsCallback(RedirectView):
+
+    allowed_cookie = True
+    allowed_header = False
+
+    def get_redirect_url(self, *args, **kwargs):
+        if not hasattr(self.request, 'session'):
+            raise ImproperlyConfigured('Django SessionMiddleware must be enabled to use MS SSO')
+
+        sign_flow = self.request.session.get('auth_token_ms_sso_auth_flow')
+        if not sign_flow:
+            messages.error(self.request, gettext('Microsoft SSO login was unsuccessful, please try it again'))
+            return '/'
+
+        result = acquire_token_by_auth_code_flow(sign_flow, self.request.GET)
+        if 'access_token' not in result:
+            messages.error(self.request, gettext('Microsoft SSO login was unsuccessful, please try it again'))
+            return sign_flow['next']
```

Comment:
Why in the condition above, you are returning `/` and here you are returning the "next" URL. Shouldn't you redirect to next URL in case of success only?

Reply:
becouse session contains next url in sign_flow. If there is no sign_flow I must redirect somewhere....

Reply:
Yes, but here, login is not sucessful - next URL typically requires sucessful login. So it wil anyway redirect back to login, or not?

Reply:
Yes this is exactly what I want. In the next login is showed flash message with messages.error text

Reply:
But you are right that this can be done without redirect